### ISSUE TITLE
change default RP duration or shard duration and data is lost

### ISSUE THREAD

**Issue by s1m0** [influxdata/influxdb#5878]
Since upgrading to 0.10.1 and realizing how much spare space I now have, I increased my 'default' retention policies from 35d (840h) to 26w (4368h) but now I can't retrieve any data that has been written since the change. 
It appears that writes are succeeding (http 204 code being returned to callers) but reads return 0 points.

**Comment by jonseymour** [comment]
@s1m0 can you see if the symptoms of #5719 apply to your case? If so, the fix will be in 0.10.2. I built a binary that contains this fix yesterday so if you want an experimental fix (supported only by me - not influxdata) I can provide you with a link to a binary on a "you have been warned, use at your own risk" basis.

**Comment by s1m0** [comment]
I don't believe it is the same - no compaction events for an hour after I altered the default retention policy on 2 of the 6 databases. 
When I checked again just now, I can read some data! I made the changes at 16:45 and by 23:33 data can be read again on one of the 2 databases, the other still refuses to deliver anything after 16:45.
If I look at the shard measurement in the _internal database it looks like data is being successfully written ("FlexCache" and "Origin" are the 2 dbs in question, "Origin" still doesn't display any data)
```
name: shard
-----------
time            clusterID               engine  fieldsCreate    hostname                id      nodeID  path                                                    seriesCreate    writePointsOk   writeReq
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  541     0       /var/opt/influxdb/data/FlexCache/default/541            224             285010          14000
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  537     0       /var/opt/influxdb/data/SASData/sixmonths/537            22              14196           1488
1456880590      9111164978299601056     tsm1    1783            vhlgnnfut001.wspan.net  555     0       /var/opt/influxdb/data/FlexCache/default/555            224             93366           4590
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  540     0       /var/opt/influxdb/data/FlexCache/default/540            240             355240          13994
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  551     0       /var/opt/influxdb/data/_internal/monitor/551            8               14800           1850
1456880590      9111164978299601056     tsm1    39              vhlgnnfut001.wspan.net  556     0       /var/opt/influxdb/data/_internal/monitor/556            11              4167            379
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  550     0       /var/opt/influxdb/data/_internal/monitor/550            9               13288           1850
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  543     0       /var/opt/influxdb/data/Origin/default/543               24              40188           6698
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  539     0       /var/opt/influxdb/data/CapacityLogger/sixmonths/539     5               22274           4456
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  538     0       /var/opt/influxdb/data/CapacityLogger/sixmonths/538     15              66846           4456
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  542     0       /var/opt/influxdb/data/Origin/default/542               24              40188           6698
1456880590      9111164978299601056     tsm1    43              vhlgnnfut001.wspan.net  557     0       /var/opt/influxdb/data/_internal/monitor/557            8               3032            379
1456880590      9111164978299601056     tsm1    0               vhlgnnfut001.wspan.net  536     0       /var/opt/influxdb/data/SASData/sixmonths/536            19              12542           1488
1456880590      9111164978299601056     tsm1    1584            vhlgnnfut001.wspan.net  554     0       /var/opt/influxdb/data/FlexCache/default/554            240             116692          4588
```

**Comment by jonseymour** [comment]
Can I ask what timezone you are in?

**Comment by s1m0** [comment]
Mountain

**Comment by jonseymour** [comment]
Ok, so the reason I asked about timezone is that if a timestamp that was actually a UTC timestamp was being interpreted as a local timestamp at the time it got inserted into Influx, then the timestamp will be dated 5 hours ahead of the actual time of that record (the difference between Mountain and GMT). Such records then wouldn't be visible in queries where that implicitly or explicitly exclude future dated events.

e.g. your loader is trying to load a record dated at 22:00 UTC (17:00 local), but some how 22:00 UTC is being interpreted at (22:00 local = 03:00 UTC). Once local time actually becomes 22:00, the record previously inserted at 22:00 UTC (17:00 local) is now no longer future dated as so appears in query results.

However, that explanation doesn't really explain why it appears in one database and not the other, except perhaps if you are running in a cluster and some of the nodes in the cluster are configured differently to others w.r.t. to timezone settings or are being loaded from clients with different timezone settings.

**Comment by s1m0** [comment]
Ah, ok, makes sense and this is a single machine, not clustered, never has been. The source data is time stamped in GMT and Grafana is set up to also run in GMT. I also was doing queries in the influx cli using relative time ("now() - 1m" for example) on the same server as influxdb - they didn't return any points either.

**Comment by jonseymour** [comment]
Yep and just so there is no confusion I should state it doesn't really matter what time zone Grafana and Influx run on - what's crucial is that if there is a parser somewhere parsing a yyyy-mm-did hh:mm:ss style time stamp and converting it into influx line protocol for purpose of writing it, that parser has to be using the time zone that matches the data.

**Comment by s1m0** [comment]
These messages were logged at the exact time that data started to be available to read from the "FlexCache" database:-
```
[tsm1wal] 2016/03/01 23:31:33 tsm1 WAL starting with 10485760 segment size
[tsm1wal] 2016/03/01 23:31:33 tsm1 WAL writing to /var/opt/influxdb/wal/FlexCache/default/555
[tsm1wal] 2016/03/01 23:31:33 tsm1 WAL starting with 10485760 segment size
[tsm1wal] 2016/03/01 23:31:33 tsm1 WAL writing to /var/opt/influxdb/wal/FlexCache/default/554
```
The "Origin" database is still not returning any points.

**Comment by s1m0** [comment]
The "Origin" database has no active shards. What do I need to do to get influxdb to create a new shard in this database?
```
name: Origin
------------
id      database        retention_policy        shard_group     start_time              end_time                expiry_time             owners
461     Origin          default                 461             2016-01-26T00:00:00Z    2016-01-27T00:00:00Z    2016-07-27T00:00:00Z    1
463     Origin          default                 463             2016-01-27T00:00:00Z    2016-01-28T00:00:00Z    2016-07-28T00:00:00Z    1
465     Origin          default                 465             2016-01-28T00:00:00Z    2016-01-29T00:00:00Z    2016-07-29T00:00:00Z    1
467     Origin          default                 467             2016-01-29T00:00:00Z    2016-01-30T00:00:00Z    2016-07-30T00:00:00Z    1
469     Origin          default                 469             2016-01-30T00:00:00Z    2016-01-31T00:00:00Z    2016-07-31T00:00:00Z    1
471     Origin          default                 471             2016-01-31T00:00:00Z    2016-02-01T00:00:00Z    2016-08-01T00:00:00Z    1
473     Origin          default                 473             2016-02-01T00:00:00Z    2016-02-02T00:00:00Z    2016-08-02T00:00:00Z    1
477     Origin          default                 477             2016-02-02T00:00:00Z    2016-02-03T00:00:00Z    2016-08-03T00:00:00Z    1
479     Origin          default                 479             2016-02-03T00:00:00Z    2016-02-04T00:00:00Z    2016-08-04T00:00:00Z    1
481     Origin          default                 481             2016-02-04T00:00:00Z    2016-02-05T00:00:00Z    2016-08-05T00:00:00Z    1
483     Origin          default                 483             2016-02-05T00:00:00Z    2016-02-06T00:00:00Z    2016-08-06T00:00:00Z    1
485     Origin          default                 485             2016-02-06T00:00:00Z    2016-02-07T00:00:00Z    2016-08-07T00:00:00Z    1
487     Origin          default                 487             2016-02-07T00:00:00Z    2016-02-08T00:00:00Z    2016-08-08T00:00:00Z    1
489     Origin          default                 489             2016-02-08T00:00:00Z    2016-02-09T00:00:00Z    2016-08-09T00:00:00Z    1
493     Origin          default                 493             2016-02-09T00:00:00Z    2016-02-10T00:00:00Z    2016-08-10T00:00:00Z    1
495     Origin          default                 495             2016-02-10T00:00:00Z    2016-02-11T00:00:00Z    2016-08-11T00:00:00Z    1
497     Origin          default                 497             2016-02-11T00:00:00Z    2016-02-12T00:00:00Z    2016-08-12T00:00:00Z    1
499     Origin          default                 499             2016-02-12T00:00:00Z    2016-02-13T00:00:00Z    2016-08-13T00:00:00Z    1
501     Origin          default                 501             2016-02-13T00:00:00Z    2016-02-14T00:00:00Z    2016-08-14T00:00:00Z    1
503     Origin          default                 503             2016-02-14T00:00:00Z    2016-02-15T00:00:00Z    2016-08-15T00:00:00Z    1
505     Origin          default                 505             2016-02-15T00:00:00Z    2016-02-16T00:00:00Z    2016-08-16T00:00:00Z    1
509     Origin          default                 509             2016-02-16T00:00:00Z    2016-02-17T00:00:00Z    2016-08-17T00:00:00Z    1
511     Origin          default                 511             2016-02-17T00:00:00Z    2016-02-18T00:00:00Z    2016-08-18T00:00:00Z    1
513     Origin          default                 513             2016-02-18T00:00:00Z    2016-02-19T00:00:00Z    2016-08-19T00:00:00Z    1
515     Origin          default                 515             2016-02-19T00:00:00Z    2016-02-20T00:00:00Z    2016-08-20T00:00:00Z    1
517     Origin          default                 517             2016-02-20T00:00:00Z    2016-02-21T00:00:00Z    2016-08-21T00:00:00Z    1
519     Origin          default                 519             2016-02-21T00:00:00Z    2016-02-22T00:00:00Z    2016-08-22T00:00:00Z    1
521     Origin          default                 521             2016-02-22T00:00:00Z    2016-02-23T00:00:00Z    2016-08-23T00:00:00Z    1
525     Origin          default                 525             2016-02-23T00:00:00Z    2016-02-24T00:00:00Z    2016-08-24T00:00:00Z    1
527     Origin          default                 527             2016-02-24T00:00:00Z    2016-02-25T00:00:00Z    2016-08-25T00:00:00Z    1
530     Origin          default                 530             2016-02-25T00:00:00Z    2016-02-26T00:00:00Z    2016-08-26T00:00:00Z    1
534     Origin          default                 534             2016-02-26T00:00:00Z    2016-02-27T00:00:00Z    2016-08-27T00:00:00Z    1
542     Origin          default                 539             2016-02-29T00:00:00Z    2016-03-01T00:00:00Z    2016-08-30T00:00:00Z    2
543     Origin          default                 539             2016-02-29T00:00:00Z    2016-03-01T00:00:00Z    2016-08-30T00:00:00Z    1
552     Origin          default                 544             2016-03-01T00:00:00Z    2016-03-02T00:00:00Z    2016-08-31T00:00:00Z    1
553     Origin          default                 544             2016-03-01T00:00:00Z    2016-03-02T00:00:00Z    2016-08-31T00:00:00Z    2
```

**Comment by jonseymour** [comment]
@s1m0  I would have thought writing points into it would ensure that there is a shard that open accepting points for whatever time period you specify in the points. Is this not the case?

**Comment by s1m0** [comment]
I've been writing points every 10 seconds for the last 8 months or more, for example, just a few seconds ago:-
```
[http] 2016/03/03 21:35:27 127.0.0.1 - - [03/Mar/2016:21:35:27 +0000] POST /write?db=Origin&precision=s&rp=default HTTP/1.1 204 0 - curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.18 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2 d86341d2-e187-11e5-b9d6-000000000000 1.500643ms
[http] 2016/03/03 21:35:29 127.0.0.1 - - [03/Mar/2016:21:35:29 +0000] POST /write?db=Origin&precision=s&rp=default HTTP/1.1 204 0 - curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.18 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2 d9c817b5-e187-11e5-b9d8-000000000000 2.761829ms
[http] 2016/03/03 21:35:29 127.0.0.1 - - [03/Mar/2016:21:35:29 +0000] POST /write?db=Origin&precision=s&rp=default HTTP/1.1 204 0 - curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.18 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2 d9c812b5-e187-11e5-b9d7-000000000000 2.924048ms
[http] 2016/03/03 21:35:29 127.0.0.1 - - [03/Mar/2016:21:35:29 +0000] POST /write?db=Origin&precision=s&rp=default HTTP/1.1 204 0 - curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.18 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2 d9d8f3b9-e187-11e5-b9d9-000000000000 117.871µs
```

**Comment by jonseymour** [comment]
That is odd - any chance you can post a gist of your server log with the query and http entries removed (to save space and also remove potentially sensitive query parameters)

**Comment by s1m0** [comment]
i removed all http and query lines but it's still 17K lines long
[https://gist.github.com/s1m0/2ebe80164abbe228b2e8#file-influxd-log](url)

**Comment by jonseymour** [comment]
Could you post a gist with the output of:

`find /var/opt/influxdb/data -ls && df /var/opt/influxdb/data`

**Comment by s1m0** [comment]
[https://gist.github.com/s1m0/ab12cf5a86363a2c235c#file-gistfile1-txt](url)

**Comment by jonseymour** [comment]
@s1m0 - thanks for that - sorry I chose the wrong directory when I originally asked. I should have asked for a find from the parent directory.

So:

`find /var/opt/influxdb -ls`

which will include the wal directories.

**Comment by s1m0** [comment]
[https://gist.github.com/s1m0/1bb8537840c44c8b413c#file-gistfile1-txt](url)

**Comment by jonseymour** [comment]
The log extracts appear to be incomplete (they don't show the server startup, for example). What time were these servers started?

**Comment by s1m0** [comment]
InfluxDB was restarted yesterday.

**Comment by jonseymour** [comment]
Ok, so I don't have a really clear idea what is going on.

My guess is that there was some problem while creating a shard covering the current period in the Origin database. There _might_ be evidence of this in the logs prior to the logs you have shared.

If you have an opportunity to restart the servers and to capture a full set of logs showing all the startup activity it might be informative. Of course, only do this if it won't be disruptive to your operations.

**Comment by s1m0** [comment]
I used "sudo kill -s SIGQUIT <PID>" then restarted the daemon and attached the logfile output

https://gist.github.com/s1m0/a0918122f160bb705e05#file-influxd-log

**Comment by jonseymour** [comment]
Thanks @s1m0. Nothing too unusual there, except for the large numbers of old ShowTagKeysMapper  goroutine calls, perhaps from an abandoned query. However, they don't appear to be blocking anything. 

There is also some evidence, I think, of a failed attempt to find other peers in a raft cluster. Was this server previously part of a raft cluster?

I take it you were still unable to write data into a new Origin shard after restarting?

Could you craft a curl command that would insert data into the Origin shard and paste a copy of the following:

- the curl command you used
- the data you used with the command
- any changes to the log resulting from execution of the command
- the listing of the  /var/opt/influxdb/ (e.g. output of `find /var/opt/influxdb -ls`)

**Comment by s1m0** [comment]
No, it has always been a single server. I have deleted the peers.json file (I think that's the right name) but it keeps getting recreated. I there a setting in the config file that would turn off attempts to contact members of a non-existent cluster?

I am continually writing to the Origin db and getting 204 returned but I can't read it!

**Comment by jonseymour** [comment]
Those raft related messages wouldn't be appearing unless some node had tried to join the cluster previously. 

One thing you might try doing is running the SHOW SERVERS command. 

If there are any servers other than the one you expect, then consider running a DROP SERVER command to drop them. As to why Origin isn't showing the results of any writes, I am afraid I am at a loss to explain this. To help further, I would need to see the outputs requested above, to confirm that the data being written is reasonable.

**Comment by s1m0** [comment]
This gets more strange...
```
> show servers
name: data_nodes
----------------
id      http_addr       tcp_addr
1       127.0.0.1:8088  127.0.0.1:8088
2       localhost:8086  localhost:8088


name: meta_nodes
----------------
id      http_addr       tcp_addr
1       localhost:8091  localhost:8088
```
And since midnight GMT data is now visible when requested from the Origin db.

show shards (partial output):
```
542     Origin          default                 539             2016-02-29T00:00:00Z    2016-03-01T00:00:00Z    2016-08-30T00:00:00Z    2
543     Origin          default                 539             2016-02-29T00:00:00Z    2016-03-01T00:00:00Z    2016-08-30T00:00:00Z    1
552     Origin          default                 544             2016-03-01T00:00:00Z    2016-03-02T00:00:00Z    2016-08-31T00:00:00Z    1
553     Origin          default                 544             2016-03-01T00:00:00Z    2016-03-02T00:00:00Z    2016-08-31T00:00:00Z    2
570     Origin          default                 553             2016-03-07T00:00:00Z    2016-03-14T00:00:00Z    2016-09-12T00:00:00Z    2
571     Origin          default                 553             2016-03-07T00:00:00Z    2016-03-14T00:00:00Z    2016-09-12T00:00:00Z    1
```
So it appears that shard creation was not working for 6 days - I'm guessing there is some logic that broke when the retention policy changed from one where a shard contained one day's worth of data to one where a shard contains 7 days' worth of data. So it needed to create a new 7-day shard (or use the existing 1-day shard for the next 6 days).

**Comment by jonseymour** [comment]
Ok, thanks for the info. I'll see if I can recreate this issue. I take it the retention policy was changed from 1 day to 7 days after you did the migration to 0.10.x?

**Comment by s1m0** [comment]
Yes, after the upgrade to 0.10.1. My data and meta are still under /var/opt rather than the new default, var/lib so I modified the .conf file before starting up v0.10.1, if that makes a difference.
Any ideas why "show servers" shows 2 data nodes on the local server?

**Comment by simcap** [comment]
I had the same issue on 0.13.0 (single instance, default configuration) while only altering (increasing) the duration of my default retention policy from 40d to 365d. 

Then we could still write points successfully but we could not query / retrieve the newly persisted data.

Since I really to keep my data longer I cannot find a way to do that without increasing the existing default retention policy.

Any fixes I am not aware about? I could raise a new issue if needed.

**Comment by lpc921** [comment]
I have the same issue on version "1.0.0". After changing the retention policy "default" from 168h to INF, data stopped coming in. It is completely silent with no error message to client or logged.

The last data logged is on line 71856 [httpd] 10.41.193.243 - logger [13/Sep/2016:15:44:45 +1200]
https://gist.github.com/lpc921/3fbe2d171217d8e48e870250dc7c341f

 It starts accepting new data again when I alter the retention policy back to 168h.

**Comment by garywgl** [comment]
I have the same issue reported by @lpc921 too. I used influx version 1.0.0-1 prebuilt deb package on Ubuntu 14.04.5. Thanks.

**Comment by beckettsean** [comment]
More reports, 1.0GA production, altered the `SHARD DURATION` and although all writes returned 204s, no data can be found or queried.

**Comment by beckettsean** [comment]
Direct inspection with `influx_inspect export` shows that the affected shards have no points written with timestamps after the `ALTER` statements. There are no other shards on disk, so the data somehow died in the write path, it appears.

**Comment by dlanderson** [comment]
Just a correction to https://github.com/influxdata/influxdb/issues/5878#issuecomment-249328263

Initial: `24h`
Changed to: `168h` and shards silently dropped data (writes in logs showing 204 response)
Fixed by changing back to `24h`

**Comment by jwilder** [comment]
@dlanderson Did the `writeDrop` stat increase when this occurred?  Can you attach the output of:

```
select derivative(sum(writeDrop),24h) from _internal.monitor."write" where time > now() - 30d group by time(24h) fill(0)
```

**Comment by dlanderson** [comment]
@jwilder Doesn't seem so:

```
> select derivative(sum(writeDrop),24h) from _internal.monitor."write" where time > now() - 30d group by time(24h) fill(0)
name: write
-----------
time                    derivative
1472256000000000000     0
1472342400000000000     0
1472428800000000000     0
1472515200000000000     0
1472601600000000000     0
1472688000000000000     0
1472774400000000000     0
1472860800000000000     0
1472947200000000000     0
1473033600000000000     0
1473120000000000000     0
1473206400000000000     0
1473292800000000000     0
1473379200000000000     0
1473465600000000000     0
1473552000000000000     0
1473638400000000000     0
1473724800000000000     0
1473811200000000000     0
1473897600000000000     0
1473984000000000000     0
1474070400000000000     0
1474156800000000000     0
1474243200000000000     0
1474329600000000000     0
1474416000000000000     0
1474502400000000000     0
1474588800000000000     0
1474675200000000000     0
1474761600000000000     0
1474848000000000000     0
```

**Comment by jwilder** [comment]
Fixed via #7370
